docs/weave/design/00-overview.md

```diff
@@ -0,0 +1,78 @@
+
+- You cannot relay on walltime (just the timestamp in the header)
+- No usage of floating point math
+- No random numbers!
+- No network calls (especially external APIs)!
+- No concurrency (unless you **really** know what you are doing)
+- JSON encoding in the storage is questionable, as the key order may change with newer JSON libraries.
+- Etc....
+
+The summary is that everything is executed sequentially and deterministically, and thus we require extremely fast transaction processing to enable high throughput. Aim for 1-2 ms per transaction, including committing to disk at the end of the block. Thus, attention to performance is very important.
+
+## ABCI
+
+To understand this design, you should first understand what an ABCI application is and how that level blockchain abstraction works. ABCI is the interface between the tendermint daemon and the state machine that processes the transactions, something akin to wsgi as the interface between apache/nginx and a django application.
+
+> There is an [in-depth reference](https://tendermint.readthedocs.io/en/master/app-development.html) to the ABCI protocol, but in short, an ABCI application is a state machine that responds to messages sent from one client (the tendermint consensus engine). It is run in parallel on every node, and they must all run the same set of transactions (what was included in the blocks), and then verify they have the same result (merkle root).
+
+The main messages that you need to be concerned with are:
+
+### Validation - CheckTx
+
+Before including a transaction, or gossiping it to peers, every node will call `CheckTx` to check if it is valid. This should be a best-attempt filter, we may reject transactions that are included in the block, but this should eliminate much spam
+
+## Execution of Blocks
+
+After a block is written to the chain, the tendermint engine makes a number of calls to process it. These are our hooks to make any *writes* to the datastore.
+
+- **BeginBlock**
+
+    BeginBlock provides the new header and block height. You can also use this as a hook to trigger any delayed tasks that will execute at a given height. (see `Ticker` below)
+
+- **DeliverTx** - once per transaction
+
+    DeliverTx is passed the raw bytes, just like CheckTx, but it is expected to process the transactions and write the state changes to the key-value store. This is the most important call to trigger any state change.
+
+- **EndBlock**
+
+    After all transactions have been processed, EndBlock is a place to communicate any configuration changes the application wishes to make on the tendermint engine. This can be changes to the validator set that signs the next block, or changes to the consensus parameters, like max block size, max numbers of transactions per block, etc.
```

Comment:
```suggestion
    After all transactions have been processed, EndBlock is a place to communicate any configuration changes the application wishes to make on the Tendermint engine. This can be changes to the validator set that signs the next block, or changes to the consensus parameters, like max block size, max numbers of transactions per block, etc.
```